lan865x/src/esp_eth_mac_lan865x.c

```diff
@@ -0,0 +1,1149 @@
+}
+
+static inline bool lan865x_spi_unlock(emac_lan865x_t *emac)
+{
+    return xSemaphoreGive(emac->spi_lock) == pdTRUE;
+}
+
+static void *lan865x_spi_init(const void *spi_config)
+{
+    void *ret = NULL;
+    ESP_GOTO_ON_FALSE(spi_config, NULL, err, TAG, "invalid spi device configuration");
+    eth_lan865x_config_t *lan865x_config = (eth_lan865x_config_t *)spi_config;
+    spi_device_interface_config_t devcfg = *(lan865x_config->spi_devcfg);
+
+    // Allocate SPI info
+    eth_spi_info_t *spi_info = NULL;
+    spi_info = heap_caps_calloc(1, sizeof(eth_spi_info_t), MALLOC_CAP_DEFAULT);
+    ESP_GOTO_ON_FALSE(spi_info, NULL, err, TAG, "no mem for SPI info");
+
+    ESP_GOTO_ON_FALSE(devcfg.command_bits == 0 && devcfg.address_bits == 0,
+                      NULL, err, TAG, "incorrect SPI frame format (command_bits/address_bits)");
+
+    // Initialize SPI device
+    ESP_GOTO_ON_FALSE(spi_bus_add_device(lan865x_config->spi_host_id, &devcfg, &spi_info->hdl) == ESP_OK,
+                      NULL, err_spi, TAG, "failed to add SPI device");
+
+    ret = spi_info;
+    return ret;
+
+err_spi:
+    free(spi_info);
+err:
+    return NULL;
+}
+
+static esp_err_t lan865x_spi_deinit(void *spi_ctx)
+{
+    esp_err_t ret = ESP_OK;
+    eth_spi_info_t *spi = (eth_spi_info_t *)spi_ctx;
+
+    spi_bus_remove_device(spi->hdl);
+    free(spi);
+    return ret;
+}
+
+static esp_err_t lan865x_spi_write(void *spi_ctx, uint32_t cmd, uint32_t addr, const void *value, uint32_t len)
+{
+    esp_err_t ret = ESP_OK;
+    eth_spi_info_t *spi = (eth_spi_info_t *)spi_ctx;
+
+    spi_transaction_t trans = {
+        .length = 8 * len,
+        .tx_buffer = value,
+    };
+
+    if (spi_device_polling_transmit(spi->hdl, &trans) != ESP_OK) {
+        ESP_LOGE(TAG, "%s(%d): spi transmit failed", __FUNCTION__, __LINE__);
+        ret = ESP_FAIL;
+    }
+
+    return ret;
+}
+
+static esp_err_t lan865x_spi_read(void *spi_ctx, uint32_t cmd, uint32_t addr, void *value, uint32_t len)
+{
+    esp_err_t ret = ESP_OK;
+    eth_spi_info_t *spi = (eth_spi_info_t *)spi_ctx;
+
+    spi_transaction_t trans = {
+        .length = 8 * len,
+        .rx_buffer = value,
+        .tx_buffer = value,
+    };
+
+    if (spi_device_polling_transmit(spi->hdl, &trans) != ESP_OK) {
+        ESP_LOGE(TAG, "%s(%d): spi transmit failed", __FUNCTION__, __LINE__);
+        ret = ESP_FAIL;
+    }
+
+    return ret;
+}
+
+static bool lan865x_parity(uint32_t value)
+{
+    uint32_t parity = 0;
+    for (int i = 1; i < 32; i++) {
+        parity ^= (value >> i) & 0x1;
+    }
+    return !(bool)parity;
+}
+
+static esp_err_t lan865x_frame_transmit(emac_lan865x_t *emac, uint8_t *frame, uint32_t length)
+{
+    esp_err_t ret = ESP_OK;
+    // the same memory is used for tx and rx blocks
+    lan865x_tx_block_t *tx_block = (lan865x_tx_block_t *)emac->spi_buffer;
+    lan865x_rx_block_t *rx_block = (lan865x_rx_block_t *)emac->spi_buffer;
+
+    if (lan865x_spi_lock(emac)) {
```

Comment:
suggest reverting the logic, as I find the combination of goto's within if-else blocks less readable
```suggestion
    if (!lan865x_spi_lock(emac)) {
        ESP_LOGE(TAG, "%s: timeout", __func__);
        return ESP_ERR_TIMEOUT;
    }
```